### PR TITLE
feat: Implement `Invite` object

### LINKED ISSUE #26: request: Implement `Invite` object
### What's the feature?

The implementation of the `invite` object

### How does it work?

_No response_

### Code of Conduct

- [X] I agree.

### PR DESCRIPTION
**Pull request by EdVraz**
This PR aims to implement the `Invite` object

resolves #26

### CODE REVIEW

File: retux/client/resources/invite.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    """A class object representing the Invite-target-types from discord."""
+
+    STREAM = 1
+    EMBEDDED_APPLICATION = 2
+
+
+@define()
+class InviteStageInstance:
+    """
+    Represents an InviteStageInstance from Discord. This is deprecated.
```
Comment (i0bs) [inline retux/client/resources/invite.py:22]:
```suggestion
    Represents an invite to a stage instance from Discord. This is deprecated.
```

File: retux/client/resources/invite.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    speaker_count: int = field()
+    "The number of users speaking in the Stage."
+    topic: str = field()
+    "The topic of the Stage instance (1-120 characters)."
+
+
+@define()
+class Invite:
+    """
+    Represents an Invite from Discord
```
Comment (i0bs) [inline retux/client/resources/invite.py:49]:
```suggestion
    Represents an invite from Discord.
```

File: retux/client/resources/invite.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    inviter : `retux.User`, optional
+        The user who created the invite, if any.
+    target_type : `InviteTargetType`, optional
+        The type of target for a voice channel invite, if any.
+    target_user : `retux.User`, optional
+        The user whose stream to display for this voice channel stream invite, if any.
+    target_application : `retux.Application`, optional
+        The embedded application to open for this voice channel embedded application invite, if any.
+    approximate_presence_count : `int`, optional
+        Approximate count of online members, returned from the `GET /invites/<code>` endpoint when `with_counts=True`.
```
Comment (i0bs) [inline retux/client/resources/invite.py:68]:
```suggestion
        Approximate count of online members, returned from the `Get Invites` endpoint
        when `with_counts=True`.
```

File: retux/client/resources/invite.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    approximate_member_count : `int`, optional
+        Approximate count of total members, returned from the `GET /invites/<code>` endpoint when `with_counts=True`.
```
Comment (i0bs) [inline retux/client/resources/invite.py:70]:
```suggestion
        Approximate count of total members, returned from the `Get Invites` endpoint
        when `with_counts=True`.
```

File: retux/client/resources/invite.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    expires_at : `datetime.datetime`, optional
+        The expiration date of this invite, returned from the `GET /invites/<code>` endpoint when `with_expiration=True`.
```
Comment (i0bs) [inline retux/client/resources/invite.py:72]:
```suggestion
        The expiration date of this invite, returned from the `Get Invites` endpoint
        when `with_expiration=True`.
```

File: retux/client/resources/invite.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    target_type: int | InviteTargetType | None = field(
+        converter=optional_c(InviteTargetType), default=None
+    )
+    """The type of target for this voice channel invite, if any."""
+    target_user: User | None = field(converter=optional_c(User), default=None)
+    """The user whose stream to display for this voice channel stream invite, if any."""
+    target_application: Application | None = field(converter=optional_c(Application), default=None)
+    """The embedded application to open for this voice channel embedded application invite, if any"""
+    approximate_presence_count: int | None = field(default=None)
+    """Approximate count of online members, returned from the `GET /invites/<code>` endpoint when `with_counts=True`."""
```
Comment (i0bs) [inline retux/client/resources/invite.py:94]:
```suggestion
    """Approximate count of online members, returned from the `Get Invites` endpoint when `with_counts=True`."""
```

File: retux/client/resources/invite.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    approximate_member_count: int | None = field(default=None)
+    """Approximate count of total members, returned from the `GET /invites/<code>` endpoint when `with_counts=True`."""
```
Comment (i0bs) [inline retux/client/resources/invite.py:96]:
```suggestion
    """Approximate count of total members, returned from the `Get Invites` endpoint when `with_counts=True`."""
```

File: retux/client/resources/invite.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    expires_at: datetime | None = field(converter=optional_c(datetime.fromisoformat), default=None)
```
Comment (i0bs) [inline retux/client/resources/invite.py:97]:
```suggestion
    expires_at: str | datetime | None = field(converter=optional_c(datetime.fromisoformat), default=None)
```

File: retux/client/resources/invite.py [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    """The channel this invite is for, if any."""
+    inviter: User | None = field(converter=optional_c(User), default=None)
+    """The user who created the invite, if any."""
+    target_type: int | InviteTargetType | None = field(
+        converter=optional_c(InviteTargetType), default=None
+    )
+    """The type of target for this voice channel invite, if any."""
+    target_user: User | None = field(converter=optional_c(User), default=None)
+    """The user whose stream to display for this voice channel stream invite, if any."""
+    target_application: Application | None = field(converter=optional_c(Application), default=None)
```
Comment (i0bs) [inline retux/client/resources/invite.py:91]:
```suggestion
    target_application: dict | Application | None = field(converter=optional_c(Application), default=None)
```

File: retux/client/resources/invite.py [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+        Approximate count of total members, returned from the `GET /invites/<code>` endpoint when `with_counts=True`.
+    expires_at : `datetime.datetime`, optional
+        The expiration date of this invite, returned from the `GET /invites/<code>` endpoint when `with_expiration=True`.
+    stage_instance : `InviteStageInstance`, optional
+        Stage instance data if there is a public Stage instance in the Stage channel this invite is for, if any. (deprecated).
+    """
+
+    code: str = field()
+    """The invite code."""
+    guild: Guild | None = field(converter=optional_c(Guild), default=None)
```
Comment (i0bs) [inline retux/client/resources/invite.py:79]:
```suggestion
    guild: dict | Guild | None = field(converter=optional_c(Guild), default=None)
```

File: retux/client/resources/invite.py [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+    """The guild this invite is for, if any."""
+    channel: Channel | None = field(converter=optional_c(Channel), default=None)
```
Comment (i0bs) [inline retux/client/resources/invite.py:81]:
```suggestion
    channel: dict | Channel | None = field(converter=optional_c(Channel), default=None)
```

File: retux/client/resources/invite.py [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+    """The channel this invite is for, if any."""
+    inviter: User | None = field(converter=optional_c(User), default=None)
```
Comment (i0bs) [inline retux/client/resources/invite.py:83]:
```suggestion
    inviter: dict | User | None = field(converter=optional_c(User), default=None)
```

File: retux/client/resources/invite.py [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+    """The user who created the invite, if any."""
+    target_type: int | InviteTargetType | None = field(
+        converter=optional_c(InviteTargetType), default=None
+    )
+    """The type of target for this voice channel invite, if any."""
+    target_user: User | None = field(converter=optional_c(User), default=None)
```
Comment (i0bs) [inline retux/client/resources/invite.py:89]:
```suggestion
    target_user: dict | User | None = field(converter=optional_c(User), default=None)
```

File: retux/client/resources/invite.py [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+    """The user whose stream to display for this voice channel stream invite, if any."""
+    target_application: Application | None = field(converter=optional_c(Application), default=None)
+    """The embedded application to open for this voice channel embedded application invite, if any"""
+    approximate_presence_count: int | None = field(default=None)
+    """Approximate count of online members, returned from the `GET /invites/<code>` endpoint when `with_counts=True`."""
+    approximate_member_count: int | None = field(default=None)
+    """Approximate count of total members, returned from the `GET /invites/<code>` endpoint when `with_counts=True`."""
+    expires_at: datetime | None = field(converter=optional_c(datetime.fromisoformat), default=None)
+    """The expiration date of this invite, returned from the `GET /invites/<code>` endpoint when `with_expiration=True`."""
+    stage_instance: InviteStageInstance | None = field(
```
Comment (i0bs) [inline retux/client/resources/invite.py:99]:
```suggestion
    stage_instance: int | InviteStageInstance | None = field(
```

File: retux/client/resources/invite.py [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+        converter=optional_c(InviteStageInstance), default=None
+    )
+    """stage instance data if there is a public Stage instance in the Stage channel this invite is for, if any. (deprecated)."""
```
Comment (i0bs) [inline retux/client/resources/invite.py:102]:
```suggestion
    """Data of the instance if there is a public stage in the channel this invite is for. (deprecated)"""
```

File: retux/client/resources/invite.py [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+    target_application : `retux.Application`, optional
+        The embedded application to open for this voice channel embedded application invite, if any.
+    approximate_presence_count : `int`, optional
+        Approximate count of online members, returned from the `GET /invites/<code>` endpoint when `with_counts=True`.
+    approximate_member_count : `int`, optional
+        Approximate count of total members, returned from the `GET /invites/<code>` endpoint when `with_counts=True`.
+    expires_at : `datetime.datetime`, optional
+        The expiration date of this invite, returned from the `GET /invites/<code>` endpoint when `with_expiration=True`.
+    stage_instance : `InviteStageInstance`, optional
+        Stage instance data if there is a public Stage instance in the Stage channel this invite is for, if any. (deprecated).
```
Comment (i0bs) [inline retux/client/resources/invite.py:74]:
```suggestion
        Data of the instance if there is a public stage in the channel this invite is for. (deprecated)
```

File: retux/client/resources/invite.py [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+    """Approximate count of total members, returned from the `GET /invites/<code>` endpoint when `with_counts=True`."""
+    expires_at: datetime | None = field(converter=optional_c(datetime.fromisoformat), default=None)
+    """The expiration date of this invite, returned from the `GET /invites/<code>` endpoint when `with_expiration=True`."""
+    stage_instance: InviteStageInstance | None = field(
+        converter=optional_c(InviteStageInstance), default=None
+    )
+    """stage instance data if there is a public Stage instance in the Stage channel this invite is for, if any. (deprecated)."""
+    # guild_scheduled_event
+    # """guild scheduled event data, only included if guild_scheduled_event_id contains a valid guild scheduled event id"""
+    # todo add this when added
```
Comment (i0bs) [inline retux/client/resources/invite.py:105]:
```suggestion
    # TODO: Implement Guild Scheduled Event object
```

**Review by i0bs — CHANGES_REQUESTED** [review]

File: retux/client/resources/invite.py [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+from ...utils.converters import optional_c, list_c
+from .application import Application
+from datetime import datetime
+
+from attrs import define, field
+from enum import IntEnum
+
+
+class InviteTargetType(IntEnum):
+    """A class object representing the Invite-target-types from discord."""
```
Comment (i0bs) [inline retux/client/resources/invite.py:13]: This needs documentation on the constants

Additionally, each constant requires a docstring of their own identical to what's written.

File: retux/client/resources/invite.py [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+    EMBEDDED_APPLICATION = 2
+
+
+@define()
+class InviteStageInstance:
+    """
+    Represents an invite to a stage instance from Discord. This is deprecated.
+
+    Attributes
+    __________
```
Comment (i0bs) [inline retux/client/resources/invite.py:25]:
```suggestion
    ----------
```

File: retux/client/resources/invite.py [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+    "The topic of the Stage instance (1-120 characters)."
+
+
+@define()
+class Invite:
+    """
+    Represents an invite from Discord.
+
+    Attributes
+    __________
```
Comment (i0bs) [inline retux/client/resources/invite.py:52]:
```suggestion
    ----------
```

File: retux/client/resources/invite.py [hunk 21]
```diff
@@ -0,0 +1,108 @@
...
+    members : `list[retux.Member]`
+        The members speaking in the Stage.
+    participant_count : `int`
+        The number of users in the Stage.
+    speaker_count : `int`
+        The number of users speaking in the Stage.
+    topic : `str`
+        The topic of the Stage instance (1-120 characters).
```
Comment (i0bs) [inline retux/client/resources/invite.py:33]:
```suggestion
        The topic of the stage instance in-between 1-120 characters.
```

File: retux/client/resources/invite.py [hunk 22]
```diff
@@ -0,0 +1,108 @@
...
+
+@define()
+class InviteStageInstance:
+    """
+    Represents an invite to a stage instance from Discord. This is deprecated.
+
+    Attributes
+    __________
+    members : `list[retux.Member]`
+        The members speaking in the Stage.
```
Comment (i0bs) [inline retux/client/resources/invite.py:27]:
```suggestion
        The members speaking in the stage.
```

File: retux/client/resources/invite.py [hunk 23]
```diff
@@ -0,0 +1,108 @@
...
+    participant_count : `int`
+        The number of users in the Stage.
```
Comment (i0bs) [inline retux/client/resources/invite.py:29]:
```suggestion
        The number of users in the stage.
```

File: retux/client/resources/invite.py [hunk 24]
```diff
@@ -0,0 +1,108 @@
...
+    speaker_count : `int`
+        The number of users speaking in the Stage.
```
Comment (i0bs) [inline retux/client/resources/invite.py:31]:
```suggestion
        The number of users speaking in the stage.
```

File: retux/client/resources/invite.py [hunk 25]
```diff
@@ -0,0 +1,108 @@
...
+    topic : `str`
+        The topic of the Stage instance (1-120 characters).
+    """
+
+    members: list[Member] = field(converter=list_c(Member))
+    """The members speaking in the Stage."""
```
Comment (i0bs) [inline retux/client/resources/invite.py:37]:
```suggestion
    """The members speaking in the stage."""
```

File: retux/client/resources/invite.py [hunk 26]
```diff
@@ -0,0 +1,108 @@
...
+    participant_count: int = field()
+    """The number of users in the Stage."""
```
Comment (i0bs) [inline retux/client/resources/invite.py:39]:
```suggestion
    """The number of users in the stage."""
```

File: retux/client/resources/invite.py [hunk 27]
```diff
@@ -0,0 +1,108 @@
...
+    speaker_count: int = field()
+    "The number of users speaking in the Stage."
```
Comment (i0bs) [inline retux/client/resources/invite.py:41]:
```suggestion
    "The number of users speaking in the stage."
```

File: retux/client/resources/invite.py [hunk 28]
```diff
@@ -0,0 +1,108 @@
...
+    topic: str = field()
+    "The topic of the Stage instance (1-120 characters)."
```
Comment (i0bs) [inline retux/client/resources/invite.py:43]:
```suggestion
    "The topic of the stage instance in-between 1-120 characters."
```

File: retux/client/resources/invite.py [hunk 29]
```diff
@@ -0,0 +1,108 @@
...
+    target_user: dict | User | None = field(converter=optional_c(User), default=None)
+    """The user whose stream to display for this voice channel stream invite, if any."""
+    target_application: dict | Application | None = field(converter=optional_c(Application), default=None)
+    """The embedded application to open for this voice channel embedded application invite, if any"""
+    approximate_presence_count: int | None = field(default=None)
+    """Approximate count of online members, returned from the `Get Invites` endpoint when `with_counts=True`."""
+    approximate_member_count: int | None = field(default=None)
+    """Approximate count of total members, returned from the `Get Invites` endpoint when `with_counts=True`."""
+    expires_at: str | datetime | None = field(converter=optional_c(datetime.fromisoformat), default=None)
+    """The expiration date of this invite, returned from the `GET /invites/<code>` endpoint when `with_expiration=True`."""
```
Comment (i0bs) [inline retux/client/resources/invite.py:101]:
```suggestion
    """The expiration date of this invite, returned from the `Get Invites` endpoint when `with_expiration=True`."""
```

**Review by i0bs — CHANGES_REQUESTED** [review]

File: retux/client/resources/invite.py (hunk 18 above)
Comment (EdVraz) [reply retux/client/resources/invite.py:13]: https://discord.com/developers/docs/resources/invite#invite-object-invite-target-types

Discord doesn't provide a description so should I just imagine smth in the context?

Comment (i0bs) [reply retux/client/resources/invite.py:13]: Yeah, pretty much, that's how a lot of the enumerables have been documented so far

File: retux/client/resources/invite.py [hunk 30]
```diff
@@ -0,0 +1,123 @@
...
+from .application import Application
+from datetime import datetime
+
+from attrs import define, field
+from enum import IntEnum
+
+
+class InviteTargetType(IntEnum):
+    """
+    A class object representing the Invite-target-types from discord.
```
Comment (i0bs) [inline retux/client/resources/invite.py:14]:
```suggestion
    Represents the targeted types of invites from Discord.
```

File: retux/client/resources/invite.py [hunk 31]
```diff
@@ -0,0 +1,123 @@
...
+
+    Constants
+    ---------
+    STREAM
+        The invite targets a stream.
+    EMBEDDED_APPLICATION
+        The invite targets  an embedded application.
```
Comment (i0bs) [inline retux/client/resources/invite.py:21]:
```suggestion
        The invite targets an embedded application.
```

**Review by i0bs — CHANGES_REQUESTED** [review]

**Review by i0bs — APPROVED** [review]
LGTM, thank you for contributing to retux!

**Pull request merged**